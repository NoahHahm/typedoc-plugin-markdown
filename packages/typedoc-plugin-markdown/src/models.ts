import { ReflectionKind, TypeDocOptionMap } from 'typedoc';

export interface TypedocPluginMarkdownOptions extends TypeDocOptionMap {
  anchorFormat: AnchorFormat;
  baseUrl: string;
  enableFrontmatter: boolean;
  entryDocument: string;
  enumMembersFormat: DataFormatStyle;
  flattenOutputFiles: boolean;
  frontmatterTags: string[];
  frontmatterGlobals: FrontmatterGlobals;
  frontmatterNamingConvention: FrontmatterNamingConvention;
  groupByKinds: boolean;
  hideBreadcrumbs: boolean;
  hideInPageTOC: boolean;
  hideKindTag: boolean;
  hidePageTitle: boolean;
  hideHierarchy: boolean;
  indentifiersAsCodeBlocks: boolean;
  indexPageTitle: string;
  kindsWithOwnFile: string | string[];
  namedAnchors: boolean;
  numberPrefixOutput: boolean;
  propertiesFormat: DataFormatStyle;
  TOCFormat: DataFormatStyle;
}

export interface TemplateMapping {
  directory: string | null;
  template: any;
  isLeaf: boolean;
  kind: ReflectionKind;
}

export type Collapse = 'object' | 'function' | 'all' | 'none';

export type DataFormatStyle = 'list' | 'table';

export type AnchorFormat = 'lowercase' | 'slug' | 'none';

export type FrontmatterGlobals =
  | string
  | Record<string, string | boolean | number | null>;

export type FrontmatterNamingConvention =
  | 'camelCase'
  | 'snakeCase'
  | 'kebabCase'
  | 'pascalCase';

export interface NavigationItem {
  title?: string;
  url?: string;
  children?: NavigationItem[];
}
