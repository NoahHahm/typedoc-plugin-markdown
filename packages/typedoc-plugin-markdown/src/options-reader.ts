import { Options, OptionsReader } from 'typedoc';
import { TypedocPluginMarkdownOptions } from './models';

export class MarkdownPluginOptionsReader implements OptionsReader {
  priority = 900;
  options: Partial<TypedocPluginMarkdownOptions>;

  name = 'custom-options';

  constructor(options: Partial<TypedocPluginMarkdownOptions>) {
    this.options = options;
  }

  read(container: Options) {
    if (this.options) {
      Object.entries(this.options).forEach(([key, value]) => {
        container.setValue(key, value);
      });
    }
  }
}
