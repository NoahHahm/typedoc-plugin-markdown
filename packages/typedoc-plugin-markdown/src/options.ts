import { Application, Options, OptionsReader, ParameterType } from 'typedoc';

export function defineOptions(app: Application) {
  app.options.addReader(
    new (class implements OptionsReader {
      name = 'markdown-theme-options-reader';
      readonly order = 900;
      readonly supportsPackages = false;
      read(container: Options) {
        if (container.getValue('theme') === 'default') {
          container.setValue('theme', 'markdown');
        }
      }
    })(),
  );

  /**
   * file output options
   */
  app.options.addDeclaration({
    name: 'entryDocument',
    help: '[Markdown Plugin] The file name of the entry document.',
    type: ParameterType.String,
    defaultValue: 'README.md',
  });

  app.options.addDeclaration({
    name: 'flattenOutput',
    help: '[Markdown Plugin] Flatten output files without directories.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'numberPrefixOutput',
    help: '[Markdown Plugin] Prefixes docs and folders by number prefixes if applicable.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'kindsWithOwnFile',
    help: "[Markdown Plugin] Specifies which reflection kinds are contained in their own file. Values 'none', 'all' OR Array of ['class', 'interface', 'enum', 'function', 'variable', 'type']",
    type: ParameterType.String | ParameterType.Array,
    defaultValue: 'all',
  });

  app.options.addDeclaration({
    name: 'groupByKinds',
    help: '[Markdown Plugin] Groups reflection kinds by headings if applicable e.g Classes, Functions. If set to false all symbols will render on the same level. Defaults to `true`',
    type: ParameterType.Boolean,
    defaultValue: true,
  });

  /**
   * ui options
   */

  app.options.addDeclaration({
    name: 'hidePageTitle',
    help: '[Markdown Plugin] Do not print page title.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideKindTag',
    help: '[Markdown Plugin] Do not print the kind tag beneath the page title.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideBreadcrumbs',
    help: '[Markdown Plugin] Do not print breadcrumbs.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideInPageTOC',
    help: '[Markdown Plugin] Do not render in-page table of contents items.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'hideHierarchy',
    help: '[Markdown Plugin] Do not print reflection hierarchy.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'indexPageTitle',
    help: '[Markdown Plugin] The title of the main index / modules page.',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    name: 'indentifiersAsCodeBlocks',
    help: '[Markdown Plugin] Format signature and declaration identifiers in code blocks.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'propertiesFormat',
    help: '[Markdown Plugin] Specify the render style of properties.',
    type: ParameterType.String,
    defaultValue: 'List',
  });

  app.options.addDeclaration({
    name: 'enumMembersFormat',
    help: '[Markdown Plugin] Specify the render style of properties.',
    type: ParameterType.String,
    defaultValue: 'List',
  });

  app.options.addDeclaration({
    name: 'typeDeclarationFormat',
    help: '[Markdown Plugin] Specify the render style of type declarations.',
    type: ParameterType.String,
    defaultValue: 'List',
  });

  /**
   * Frontmatter options
   */
  app.options.addDeclaration({
    name: 'enableFrontmatter',
    help: '[Markdown Plugin] Prepend frontmatter to output.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'frontmatterTags',
    help: '[Markdown Plugin] Specify which file comment tags should be added to frontmatter.',
    type: ParameterType.Array,
    defaultValue: [],
  });

  app.options.addDeclaration({
    name: 'frontmatterGlobals',
    help: '[Markdown Plugin] Specify static variables to be added to all frontmatter.',
    type: ParameterType.Mixed,
    defaultValue: {},
  });

  app.options.addDeclaration({
    name: 'frontmatterNamingConvention',
    help: '[Markdown Plugin] Specify the naming convention of front matter variables.',
    type: ParameterType.String,
    defaultValue: 'camelCase',
    validate: (option) => {
      const availableValues = [
        'camelCase',
        'snakeCase',
        'kebabCase',
        'pascalCase',
      ];
      if (!availableValues.includes(option)) {
        throw new Error(
          `Unexpected value for frontmatterNamingConvention, the expected value is one of 'camelCase', 'snakeCase','kebabCase','pascalCase'`,
        );
      }
    },
  });

  /**
   * Utility options
   */
  app.options.addDeclaration({
    help: '[Markdown Plugin] Specifies the base url for internal link. If omitted all urls will be relative.',
    name: 'baseUrl',
    type: ParameterType.String,
  });

  app.options.addDeclaration({
    name: 'namedAnchors',
    help: '[Markdown Plugin] Use HTML named anchors as fragment identifiers for engines that do not automatically assign header ids. Should be set for Bitbucket Server docs.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });

  app.options.addDeclaration({
    name: 'preserveAnchorCasing',
    help: '[Markdown Plugin] Preserve anchor casing when generating links.',
    type: ParameterType.Boolean,
    defaultValue: false,
  });
}
