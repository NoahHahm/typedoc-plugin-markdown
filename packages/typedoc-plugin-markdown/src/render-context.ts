import * as path from 'path';
import { Options, PageEvent, Reflection } from 'typedoc';
import { URL_PREFIX } from './support/constants';

/* start_imports */
import { breadcrumbs } from './resources/partials/breadcrumbs';
import { commentParts } from './resources/partials/comment.parts';
import { comment } from './resources/partials/comment';
import { footer } from './resources/partials/footer';
import { header } from './resources/partials/header';
import { hierarchy } from './resources/partials/hierarchy';
import { declarationMemberIdentifier } from './resources/partials/member.declaration.identifier';
import { declarationMemberName } from './resources/partials/member.declaration.name';
import { declarationMember } from './resources/partials/member.declaration';
import { memberIndex } from './resources/partials/member.index';
import { indexSignatureTitle } from './resources/partials/member.indexsignature.title';
import { inheritance } from './resources/partials/member.inheritance';
import { memberKindTag } from './resources/partials/member.kind-tag';
import { referenceMember } from './resources/partials/member.reference';
import { reflectionMember } from './resources/partials/member.reflection';
import { signatureMemberIdentifier } from './resources/partials/member.signature.identifier';
import { signatureMember } from './resources/partials/member.signature';
import { sources } from './resources/partials/member.sources';
import { memberTitle } from './resources/partials/member.title';
import { member } from './resources/partials/member';
import { typeDeclarationMember } from './resources/partials/member.type-declaration';
import { members } from './resources/partials/members';
import { navigation } from './resources/partials/navigation';
import { pageIndex } from './resources/partials/page.index';
import { pageTitle } from './resources/partials/page.title';
import { enumMembersTable } from './resources/partials/table.enum-members';
import { parametersTable } from './resources/partials/table.parameters';
import { propertiesTable } from './resources/partials/table.properties';
import { typeParametersTable } from './resources/partials/table.typeparameters';
import { arrayType } from './resources/partials/type.array';
import { conditionalType } from './resources/partials/type.conditional';
import { declarationType } from './resources/partials/type.declaration';
import { functionType } from './resources/partials/type.function';
import { indexAccessType } from './resources/partials/type.index-access';
import { inferredType } from './resources/partials/type.inferred';
import { intersectionType } from './resources/partials/type.intersection';
import { intrinsicType } from './resources/partials/type.intrinsic';
import { literalType } from './resources/partials/type.literal';
import { queryType } from './resources/partials/type.query';
import { referenceType } from './resources/partials/type.reference';
import { reflectionType } from './resources/partials/type.reflection';
import { someType } from './resources/partials/type.some';
import { tupleType } from './resources/partials/type.tuple';
import { typeOperatorType } from './resources/partials/type.type-operator';
import { unionType } from './resources/partials/type.union';
import { unknownType } from './resources/partials/type.unknown';
import { memberTemplate } from './resources/templates/member';
import { projectTemplate } from './resources/templates/project';
import { readmeTemplate } from './resources/templates/read-me';
import { reflectionTemplate } from './resources/templates/reflection';
/* end_imports */

function bind<F, L extends any[], R>(fn: (f: F, ...a: L) => R, first: F) {
  return (...r: L) => fn(first, ...r);
}

export class MarkdownThemeRenderContext {
  constructor(
    public page: PageEvent<Reflection> | null,
    public options: Options,
  ) {}

  urlTo = (reflection: Reflection) => {
    return this.relativeURL(reflection.url);
  };

  relativeURL = (url: string | undefined) => {
    if (!url) {
      return null;
    }
    if (URL_PREFIX.test(url)) {
      return url;
    } else {
      if (this.options.getValue('baseUrl')) {
        return this.options.getValue('baseUrl') + url.replace(/\\/g, '/');
      }

      const relative = path.relative(
        path.dirname(this.page?.url || '.'),
        path.dirname(url),
      );

      return this.parseUrl(
        path.join(relative, path.basename(url)).replace(/\\/g, '/'),
      );
    }
  };

  parseUrl(url: string) {
    return encodeURI(url);
  }

  /* start_resources */

  // templates
  memberTemplate = bind(memberTemplate, this);
  projectTemplate = bind(projectTemplate, this);
  readmeTemplate = bind(readmeTemplate, this);
  reflectionTemplate = bind(reflectionTemplate, this);

  // partials
  breadcrumbs = bind(breadcrumbs, this);
  commentParts = bind(commentParts, this);
  comment = bind(comment, this);
  footer = bind(footer, this);
  header = bind(header, this);
  hierarchy = bind(hierarchy, this);
  declarationMemberIdentifier = bind(declarationMemberIdentifier, this);
  declarationMemberName = bind(declarationMemberName, this);
  declarationMember = bind(declarationMember, this);
  memberIndex = bind(memberIndex, this);
  indexSignatureTitle = bind(indexSignatureTitle, this);
  inheritance = bind(inheritance, this);
  memberKindTag = bind(memberKindTag, this);
  referenceMember = bind(referenceMember, this);
  reflectionMember = bind(reflectionMember, this);
  signatureMemberIdentifier = bind(signatureMemberIdentifier, this);
  signatureMember = bind(signatureMember, this);
  sources = bind(sources, this);
  memberTitle = bind(memberTitle, this);
  member = bind(member, this);
  typeDeclarationMember = bind(typeDeclarationMember, this);
  members = bind(members, this);
  navigation = bind(navigation, this);
  pageIndex = bind(pageIndex, this);
  pageTitle = bind(pageTitle, this);
  enumMembersTable = bind(enumMembersTable, this);
  parametersTable = bind(parametersTable, this);
  propertiesTable = bind(propertiesTable, this);
  typeParametersTable = bind(typeParametersTable, this);
  arrayType = bind(arrayType, this);
  conditionalType = bind(conditionalType, this);
  declarationType = bind(declarationType, this);
  functionType = bind(functionType, this);
  indexAccessType = bind(indexAccessType, this);
  inferredType = bind(inferredType, this);
  intersectionType = bind(intersectionType, this);
  intrinsicType = bind(intrinsicType, this);
  literalType = bind(literalType, this);
  queryType = bind(queryType, this);
  referenceType = bind(referenceType, this);
  reflectionType = bind(reflectionType, this);
  someType = bind(someType, this);
  tupleType = bind(tupleType, this);
  typeOperatorType = bind(typeOperatorType, this);
  unionType = bind(unionType, this);
  unknownType = bind(unknownType, this);
  /* end_resources */
}
