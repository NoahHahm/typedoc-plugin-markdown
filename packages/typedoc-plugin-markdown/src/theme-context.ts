import * as path from 'path';
import { DeclarationReflection, Options, PageEvent, Reflection } from 'typedoc';
import { TypedocPluginMarkdownOptions } from './models';
import { partials, templates } from './resources';
import { URL_PREFIX } from './support/constants';
import { getReflectionTitle } from './support/helpers';
import { unEscapeChars } from './support/utils';
import { MarkdownTheme } from './theme';

export class MarkdownThemeRenderContext {
  public globalsFile = 'modules.md';

  private _activeLocation: string;

  set activeLocation(activeLocation: string) {
    this._activeLocation = activeLocation;
  }

  get activeLocation() {
    return this._activeLocation;
  }

  constructor(public theme: MarkdownTheme, public options: Options) {}

  getOption<K extends keyof TypedocPluginMarkdownOptions>(name: K) {
    return this.options.getValue(name) as TypedocPluginMarkdownOptions[K];
  }

  getRelativeUrl(url: string | undefined) {
    if (!url) {
      return null;
    }
    if (URL_PREFIX.test(url)) {
      return url;
    } else {
      if (this.getOption('baseUrl')) {
        return this.getOption('baseUrl') + url.replace(/\\/g, '/');
      }

      const relative = path.relative(
        path.dirname(this.activeLocation),
        path.dirname(url),
      );

      return path.join(relative, path.basename(url)).replace(/\\/g, '/');
    }
  }

  getBaseFrontmatterVars(page: PageEvent<Reflection>) {
    return {
      title: unEscapeChars(
        getReflectionTitle(page.model as DeclarationReflection),
      ),
    };
  }

  urlTo = (reflection: Reflection) => this.relativeURL(reflection.url);

  relativeURL = (url: string | undefined) => {
    return this.getRelativeUrl(url);
  };

  baseFrontmatterVars = (page: PageEvent<Reflection>) => {
    return this.getBaseFrontmatterVars(page);
  };

  templates = templates(this);
  partials = partials(this);
}
