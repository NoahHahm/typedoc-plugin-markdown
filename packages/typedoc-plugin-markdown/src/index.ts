import { Application } from 'typedoc';
import { defineOptions } from './options';
import { generateMarkdown, renderMarkdown } from './renderer';
import { MarkdownTheme } from './theme';
import { MarkdownThemeRenderContext } from './theme-context';

export function load(app: Application) {
  /**
   * Exposes markdown theme to the renderer
   */
  app.renderer.defineTheme('markdown', MarkdownTheme);

  /**
   * Defines all plugin options
   */
  defineOptions(app);

  /**
   * Decouple HTML logic from the renderer (there should probably be a better solution to this)
   */
  Object.defineProperty(app, 'generateDocs', { value: generateMarkdown });
  Object.defineProperty(app.renderer, 'render', {
    value: renderMarkdown,
    configurable: true,
  });
}

/**
 * Expose global entrypoints
 */
export * from './models';
export * from './options-reader';
export { partials } from './resources';
export { MarkdownTheme, MarkdownThemeRenderContext };
