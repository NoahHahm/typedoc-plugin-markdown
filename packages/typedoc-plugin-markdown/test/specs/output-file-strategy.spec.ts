import { ProjectReflection } from 'typedoc';
import { MarkdownTheme } from '../../src';

describe(`Output File Strategy`, () => {
  let project: ProjectReflection;
  let theme: MarkdownTheme;

  test(`should set default --kindsWithOwnFile 'all' (modules)`, async () => {
    ({ project, theme } = await global.bootstrap(
      ['modules/module-1', 'modules/module-2'],
      { options: { kindsWithOwnFile: 'all' } },
    ));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
    expect(theme.getNavigation(project)).toMatchSnapshot();
  });

  test(`should set default --kindsWithOwnFile 'all' (modules) for monorepo`, async () => {
    ({ project, theme } = await global.bootstrap(
      ['modules/module-1', 'modules/module-2'],
      { options: { kindsWithOwnFile: 'all' } },
    ));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
    expect(theme.getNavigation(project)).toMatchSnapshot();
  });

  test(`should set default --kindsWithOwnFile 'all' (namespaces)`, async () => {
    ({ project, theme } = await global.bootstrap(['index.ts'], {
      options: { kindsWithOwnFile: 'all' },
    }));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
  });

  test(`should set --kindsWithOwnFile 'all' --numberPrefixOutput`, async () => {
    ({ project, theme } = await global.bootstrap(
      ['modules/module-1', 'modules/module-2'],
      { options: { kindsWithOwnFile: 'all', numberPrefixOutput: true } },
    ));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
  });

  test(`should set --kindsWithOwnFile 'all' --flattenOutputFiles`, async () => {
    ({ project, theme } = await global.bootstrap(
      ['modules/module-1', 'modules/module-2'],
      { options: { kindsWithOwnFile: 'all', flattenOutputFiles: true } },
    ));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
  });

  test(`should set with --groupByKinds=false (modules)`, async () => {
    ({ project, theme } = await global.bootstrap(
      ['modules/module-1', 'modules/module-2'],
      { options: { groupByKinds: false } },
    ));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
  });

  test(`should set --kindsWithOwnFile none'`, async () => {
    ({ project, theme } = await global.bootstrap(
      ['modules/module-1', 'modules/module-2'],
      { options: { kindsWithOwnFile: 'none' } },
    ));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
    expect(theme.getNavigation(project)).toMatchSnapshot();
  });

  test(`should set default --kindsWithOwnFile 'none' (namespaces)`, async () => {
    ({ project, theme } = await global.bootstrap(['index.ts'], {
      options: { kindsWithOwnFile: 'none' },
    }));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
  });

  test(`should --kindsWithOwnFile [class]'`, async () => {
    ({ project, theme } = await global.bootstrap(
      ['modules/module-1', 'modules/module-2'],
      {
        options: { kindsWithOwnFile: ['Class'] },
      },
    ));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
  });

  test(`should --kindsWithOwnFile [Class, Interface]'`, async () => {
    ({ project, theme } = await global.bootstrap(
      ['modules/module-1', 'modules/module-2'],
      {
        options: {
          kindsWithOwnFile: ['class', 'interface'],
        },
      },
    ));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
  });

  test(`should set kindsWithOwnFile [enum, types, variables]'`, async () => {
    ({ project, theme } = await global.bootstrap(
      ['modules/module-1', 'modules/module-2'],
      {
        options: {
          kindsWithOwnFile: ['Enum', 'TypeAlias', 'Variable'],
        },
      },
    ));
    expect(theme.getUrls(project).map((url) => url.url)).toMatchSnapshot();
  });
});
