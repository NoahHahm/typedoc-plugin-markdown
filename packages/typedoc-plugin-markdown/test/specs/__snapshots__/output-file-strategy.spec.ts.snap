// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output File Strategy should --kindsWithOwnFile [Class, Interface]' 1`] = `
[
  "modules.md",
  "README.md",
  "module-1/module-1.md",
  "module-1/namespaces/someNameSpace/namespace.someNameSpace.md",
  "module-1/namespaces/someNameSpace/namespaces/nestedNamespace/namespace.nestedNamespace.md",
  "module-1/namespaces/someNameSpace/classes/class.NameSpaceClassA.md",
  "module-1/namespaces/someNameSpace/interfaces/interface.NamespaceInterface.md",
  "module-1/classes/class.Module1ClassA.md",
  "module-1/classes/class.Module1ClassB.md",
  "module-1/interfaces/interface.Module1InterfaceA.md",
  "module-1/interfaces/interface.Module1InterfaceB.md",
  "module-2/module-2.md",
  "module-2/classes/class.Module2ClassA.md",
  "module-2/classes/class.Module2ClassB.md",
  "module-2/interfaces/interface.Module2InterfaceA.md",
  "module-2/interfaces/interface.Module2InterfaceB.md",
]
`;

exports[`Output File Strategy should --kindsWithOwnFile [class]' 1`] = `
[
  "modules.md",
  "README.md",
  "module-1/module-1.md",
  "module-1/namespaces/someNameSpace/namespace.someNameSpace.md",
  "module-1/namespaces/someNameSpace/namespaces/nestedNamespace/namespace.nestedNamespace.md",
  "module-1/namespaces/someNameSpace/classes/class.NameSpaceClassA.md",
  "module-1/classes/class.Module1ClassA.md",
  "module-1/classes/class.Module1ClassB.md",
  "module-2/module-2.md",
  "module-2/classes/class.Module2ClassA.md",
  "module-2/classes/class.Module2ClassB.md",
]
`;

exports[`Output File Strategy should set --kindsWithOwnFile none' 1`] = `
[
  "modules.md",
  "README.md",
  "module-1/module-1.md",
  "module-1/namespaces/someNameSpace/namespace.someNameSpace.md",
  "module-1/namespaces/someNameSpace/namespaces/namespace.nestedNamespace.md",
  "module-2.md",
]
`;

exports[`Output File Strategy should set default --kindsWithOwnFile 'all' (modules) 1`] = `
[
  "modules.md",
  "README.md",
  "module-1/module-1.md",
  "module-1/namespaces/someNameSpace/namespace.someNameSpace.md",
  "module-1/namespaces/someNameSpace/namespaces/nestedNamespace/namespace.nestedNamespace.md",
  "module-1/namespaces/someNameSpace/namespaces/nestedNamespace/variables/variable.nestedNamespaceVariable.md",
  "module-1/namespaces/someNameSpace/enums/enumeration.NamespaceEnum.md",
  "module-1/namespaces/someNameSpace/classes/class.NameSpaceClassA.md",
  "module-1/namespaces/someNameSpace/interfaces/interface.NamespaceInterface.md",
  "module-1/namespaces/someNameSpace/types/type-alias.NamespaceTypeLiteral.md",
  "module-1/namespaces/someNameSpace/variables/variable.namespaceVariable.md",
  "module-1/enums/enumeration.Module1EnumA.md",
  "module-1/enums/enumeration.Module1EnumB.md",
  "module-1/classes/class.Module1ClassA.md",
  "module-1/classes/class.Module1ClassB.md",
  "module-1/interfaces/interface.Module1InterfaceA.md",
  "module-1/interfaces/interface.Module1InterfaceB.md",
  "module-1/types/type-alias.Module1TypeLiteralA.md",
  "module-1/types/type-alias.Module1TypeLiteralB.md",
  "module-1/variables/variable.module1VariableA.md",
  "module-1/variables/variable.module1VariableB.md",
  "module-1/functions/function.module1FunctionA.md",
  "module-1/functions/function.module1FunctionB.md",
  "module-2/module-2.md",
  "module-2/enums/enumeration.Module2EnumA.md",
  "module-2/enums/enumeration.Module2EnumB.md",
  "module-2/classes/class.Module2ClassA.md",
  "module-2/classes/class.Module2ClassB.md",
  "module-2/interfaces/interface.Module2InterfaceA.md",
  "module-2/interfaces/interface.Module2InterfaceB.md",
  "module-2/types/type-alias.Module2TypeLiteralA.md",
  "module-2/types/type-alias.Module2TypeLiteralB.md",
  "module-2/variables/variable.module2VariableA.md",
  "module-2/variables/variable.module2VariableB.md",
  "module-2/functions/function.module2FunctionA.md",
  "module-2/functions/function.module2sFunctionB.md",
]
`;

exports[`Output File Strategy should set default --kindsWithOwnFile 'all' (namespaces) 1`] = `
[
  "modules.md",
  "README.md",
  "namespaces/someParentNamespace/namespace.someParentNamespace.md",
  "namespaces/someParentNamespace/namespaces/someNameSpace/namespace.someNameSpace.md",
  "namespaces/someParentNamespace/namespaces/someNameSpace/namespaces/nestedNamespace/namespace.nestedNamespace.md",
  "namespaces/someParentNamespace/namespaces/someNameSpace/namespaces/nestedNamespace/variables/variable.nestedNamespaceVariable.md",
  "namespaces/someParentNamespace/namespaces/someNameSpace/enums/enumeration.NamespaceEnum.md",
  "namespaces/someParentNamespace/namespaces/someNameSpace/classes/class.NameSpaceClassA.md",
  "namespaces/someParentNamespace/namespaces/someNameSpace/interfaces/interface.NamespaceInterface.md",
  "namespaces/someParentNamespace/namespaces/someNameSpace/types/type-alias.NamespaceTypeLiteral.md",
  "namespaces/someParentNamespace/namespaces/someNameSpace/variables/variable.namespaceVariable.md",
  "namespaces/someParentNamespace/enums/enumeration.Module1EnumA.md",
  "namespaces/someParentNamespace/enums/enumeration.Module1EnumB.md",
  "namespaces/someParentNamespace/classes/class.Module1ClassA.md",
  "namespaces/someParentNamespace/classes/class.Module1ClassB.md",
  "namespaces/someParentNamespace/interfaces/interface.Module1InterfaceA.md",
  "namespaces/someParentNamespace/interfaces/interface.Module1InterfaceB.md",
  "namespaces/someParentNamespace/types/type-alias.Module1TypeLiteralA.md",
  "namespaces/someParentNamespace/types/type-alias.Module1TypeLiteralB.md",
  "namespaces/someParentNamespace/variables/variable.module1VariableA.md",
  "namespaces/someParentNamespace/variables/variable.module1VariableB.md",
  "namespaces/someParentNamespace/functions/function.module1FunctionA.md",
  "namespaces/someParentNamespace/functions/function.module1FunctionB.md",
  "enums/enumeration.EnumDeclarations.md",
  "enums/enumeration.EnumDeclarationsWithDefaults.md",
  "classes/class.Animal.md",
  "classes/class.Base.md",
  "classes/class.C.md",
  "classes/class.Derived.md",
  "classes/class.Dog.md",
  "classes/class.IndexableClass.md",
  "classes/class.Point.md",
  "types/type-alias.AnyFunctionType.md",
  "variables/variable.__DOUBLE_UNDERSCORES_DECLARATION__.md",
  "variables/variable.enuRef.md",
  "variables/variable.getterAndSetter.md",
  "variables/variable.indexableDeclaration.md",
  "variables/variable.objectLiteralAsConstDeclaration.md",
  "variables/variable.objectLiteralDeclaration.md",
  "variables/variable.stringConstWithDefaultValue.md",
  "variables/variable.stringLetWithDefaultValue.md",
  "variables/variable.typeLiteralDeclaration.md",
  "variables/variable.undefinedNumberDeclaration.md",
  "functions/function.functionDeclaration.md",
]
`;

exports[`Output File Strategy should set default --kindsWithOwnFile 'none' (namespaces) 1`] = `
[
  "modules.md",
  "README.md",
  "namespaces/someParentNamespace/namespace.someParentNamespace.md",
  "namespaces/someParentNamespace/namespaces/someNameSpace/namespace.someNameSpace.md",
  "namespaces/someParentNamespace/namespaces/someNameSpace/namespaces/namespace.nestedNamespace.md",
]
`;

exports[`Output File Strategy should set kindsWithOwnFile [enum, types, variables]' 1`] = `
[
  "modules.md",
  "README.md",
  "module-1/module-1.md",
  "module-1/namespaces/someNameSpace/namespace.someNameSpace.md",
  "module-1/namespaces/someNameSpace/namespaces/nestedNamespace/namespace.nestedNamespace.md",
  "module-1/namespaces/someNameSpace/namespaces/nestedNamespace/variables/variable.nestedNamespaceVariable.md",
  "module-1/namespaces/someNameSpace/enums/enumeration.NamespaceEnum.md",
  "module-1/namespaces/someNameSpace/types/type-alias.NamespaceTypeLiteral.md",
  "module-1/namespaces/someNameSpace/variables/variable.namespaceVariable.md",
  "module-1/enums/enumeration.Module1EnumA.md",
  "module-1/enums/enumeration.Module1EnumB.md",
  "module-1/types/type-alias.Module1TypeLiteralA.md",
  "module-1/types/type-alias.Module1TypeLiteralB.md",
  "module-1/variables/variable.module1VariableA.md",
  "module-1/variables/variable.module1VariableB.md",
  "module-2/module-2.md",
  "module-2/enums/enumeration.Module2EnumA.md",
  "module-2/enums/enumeration.Module2EnumB.md",
  "module-2/types/type-alias.Module2TypeLiteralA.md",
  "module-2/types/type-alias.Module2TypeLiteralB.md",
  "module-2/variables/variable.module2VariableA.md",
  "module-2/variables/variable.module2VariableB.md",
]
`;

exports[`Output File Strategy should set with --groupByKinds=false (modules) 1`] = `
[
  "modules.md",
  "README.md",
  "module-1/module-1.md",
  "module-1/namespaces/someNameSpace/namespace.someNameSpace.md",
  "module-1/namespaces/someNameSpace/namespaces/nestedNamespace/namespace.nestedNamespace.md",
  "module-1/namespaces/someNameSpace/namespaces/nestedNamespace/variable.nestedNamespaceVariable.md",
  "module-1/namespaces/someNameSpace/enumeration.NamespaceEnum.md",
  "module-1/namespaces/someNameSpace/class.NameSpaceClassA.md",
  "module-1/namespaces/someNameSpace/interface.NamespaceInterface.md",
  "module-1/namespaces/someNameSpace/type-alias.NamespaceTypeLiteral.md",
  "module-1/namespaces/someNameSpace/variable.namespaceVariable.md",
  "module-1/enumeration.Module1EnumA.md",
  "module-1/enumeration.Module1EnumB.md",
  "module-1/class.Module1ClassA.md",
  "module-1/class.Module1ClassB.md",
  "module-1/interface.Module1InterfaceA.md",
  "module-1/interface.Module1InterfaceB.md",
  "module-1/type-alias.Module1TypeLiteralA.md",
  "module-1/type-alias.Module1TypeLiteralB.md",
  "module-1/variable.module1VariableA.md",
  "module-1/variable.module1VariableB.md",
  "module-1/function.module1FunctionA.md",
  "module-1/function.module1FunctionB.md",
  "module-2/module-2.md",
  "module-2/enumeration.Module2EnumA.md",
  "module-2/enumeration.Module2EnumB.md",
  "module-2/class.Module2ClassA.md",
  "module-2/class.Module2ClassB.md",
  "module-2/interface.Module2InterfaceA.md",
  "module-2/interface.Module2InterfaceB.md",
  "module-2/type-alias.Module2TypeLiteralA.md",
  "module-2/type-alias.Module2TypeLiteralB.md",
  "module-2/variable.module2VariableA.md",
  "module-2/variable.module2VariableB.md",
  "module-2/function.module2FunctionA.md",
  "module-2/function.module2sFunctionB.md",
]
`;
